drake/solvers/Optimization.h

```diff
@@ -246,20 +278,27 @@ namespace Drake {
 //    const DecisionVariable& addIntegerVariables(size_t num_new_vars, std::string name);
 //  ...
 
-    void addCost(const std::shared_ptr<Constraint>& obj) {
-      checkVariables(obj);
+    void addCost(std::shared_ptr<Constraint> const& obj, VariableList const& vars) {
```

Comment:
Do you mean that `addLinearConstraint` should not be its own method and that instead it would be an `addConstraint` overload that infers a linear nature based on its argument types?  If we add more metadata to FunctionTraits then we can have a function structure declare itself as generic, linear, or linear-equality.